support for multibuild / locallink feature of OBS

lethliel opened the pull request:
Adds multibuild output to the osc results output. This feature is needed to give the user the possibility to show all results. 

The normal output looks like this: 

```
# osc r 
openSUSE_Tumbleweed  i586       excluded
openSUSE_Tumbleweed  x86_64     excluded
images               x86_64     excluded
images               m68k       excluded
images               ppc64le    excluded
OSU_SLE_12           m68k       excluded
OSU_SLE_12           x86_64     excluded
```

You can now add a -m (--multi-package) or a -s (--subpackage) switch to get the results of the multibuild packages too.

1. osc r -m 
get the result from main package and all sub packages
2. osc r -s SUBPACK
get only the result from SUBPACK


**Example osc r -m:**
```
# osc r -m 
openSUSE_Tumbleweed  i586       kernel                                   excluded
openSUSE_Tumbleweed  i586       kernel:kernel-obs-build                  failed
openSUSE_Tumbleweed  i586       kernel:kernel-source                     succeeded
openSUSE_Tumbleweed  x86_64     kernel                                   excluded
openSUSE_Tumbleweed  x86_64     kernel:kernel-obs-build                  failed
openSUSE_Tumbleweed  x86_64     kernel:kernel-source                     succeeded
images               x86_64     kernel                                   excluded
images               x86_64     kernel:kernel-obs-build                  excluded
images               x86_64     kernel:kernel-source                     excluded
images               m68k       kernel                                   excluded
images               m68k       kernel:kernel-obs-build                  excluded
images               m68k       kernel:kernel-source                     excluded
images               ppc64le    kernel                                   excluded
images               ppc64le    kernel:kernel-obs-build                  excluded
images               ppc64le    kernel:kernel-source                     excluded
OSU_SLE_12           m68k       kernel                                   excluded
OSU_SLE_12           m68k       kernel:kernel-obs-build                  excluded
OSU_SLE_12           m68k       kernel:kernel-source                     unresolvable
OSU_SLE_12           x86_64     kernel                                   excluded
OSU_SLE_12           x86_64     kernel:kernel-obs-build                  failed
OSU_SLE_12           x86_64     kernel:kernel-source                     succeeded
```

**Example osc r -s**

```
# osc r -s kernel-source
openSUSE_Tumbleweed  i586       kernel:kernel-source                     succeeded
openSUSE_Tumbleweed  x86_64     kernel:kernel-source                     succeeded
images               x86_64     kernel:kernel-source                     excluded
images               m68k       kernel:kernel-source                     excluded
images               ppc64le    kernel:kernel-source                     excluded
OSU_SLE_12           m68k       kernel:kernel-source                     unresolvable
OSU_SLE_12           x86_64     kernel:kernel-source                     succeeded
```

marcus-h commented:
On 2016-11-25 06:00:18 -0800, Marco Strigl wrote:
> Adds multibuild output to the osc results output. This feature is needed to give the user the possibility to show all results. 
> 
I just had a very very quick glance at the patch:
- the indention is screwed up
- the global module "multicount" variable looks really fishy;)

I'll have a closer look at this later (Also, I need to have a
look at the general multibuild concept first).

> **Example osc r -M**
> 
> ```
> # osc r -M kernel-source

Hmm probably a dumb question, but why do we need this? Can't we just
run "osc results kernel-source"? (btw., is CWD an osc project working
copy or a package working copy?)

adrianschroeter commented:
On Freitag, 25. November 2016, 07:26:18 CET wrote Marcus H�we:
> On 2016-11-25 06:00:18 -0800, Marco Strigl wrote:
...
> > **Example osc r -M**
> > 
> > ```
> > # osc r -M kernel-source
> 
> Hmm probably a dumb question, but why do we need this? Can't we just
> run "osc results kernel-source"? (btw., is CWD an osc project working
> copy or a package working copy?)

Current osc interface would handle kernel-source as a project...

But 

 osc r $YOUR_project kernel:kernel-source

would work (assuming kernel "kernel" as main container and "kernel-source.spec").

You can find an example setup here:

 https://build-test.opensuse.org/project/monitor/home:adrianSuSE

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imend�rffer, Jane Smithard, Graham Norton, HRB 21284 (AG N�rnberg) 
 
Maxfeldstra�e 5                         
90409 N�rnberg 
Germany

marcus-h commented:
On 2016-11-25 07:43:27 -0800, Adrian Schröter wrote:
> On Freitag, 25. November 2016, 07:26:18 CET wrote Marcus H�we:
> > On 2016-11-25 06:00:18 -0800, Marco Strigl wrote:
> ...
> > > **Example osc r -M**
> > > 
> > > ```
> > > # osc r -M kernel-source
> > 
> > Hmm probably a dumb question, but why do we need this? Can't we just
> > run "osc results kernel-source"? (btw., is CWD an osc project working
> > copy or a package working copy?)
> 
> Current osc interface would handle kernel-source as a project...
> 
Arghs... yes... *sigh*

> But 
> 
>  osc r $YOUR_project kernel:kernel-source
> 
> would work (assuming kernel "kernel" as main container and "kernel-source.spec").
> 
> You can find an example setup here:
> 
>  https://build-test.opensuse.org/project/monitor/home:adrianSuSE
> 
Great, thanks!

Btw., https://build-test.opensuse.org states "Every openSUSE user can login",
but it doesn't work for me:) (I'm getting a 401 using the same credentials
as for api.o.o) (but the /public/ route is enough for this;) ).

lethliel commented:
First of all @marcus-h: Thanks for the review and the comments.

> I just had a very very quick glance at the patch:
>
>    
> - the indention is screwed up

I thought I payed attention to indention. When I run osc it works. Can you tell me what is screwed?
I have a knack for screwing indentions.

> - the global module "multicount" variable looks really fishy;)

I don't like it either. But I could not think of a better way. If you have an idea I would be thankful.

marcus-h commented:
On 2016-11-25 11:07:55 -0800, Marco Strigl wrote:
> First of all @marcus-h: Thanks for the review and the comments.
> 
> > I just had a very very quick glance at the patch:
> >
> >    
> > - the indention is screwed up
> 
> I thought I payed attention to indention. When I run osc it works. Can you tell me what is screwed?

You sometimes indented with 2 spaces and sometimes with 4 spaces.
The default (in the osc codebase) is 4 spaces.

> I have a knack for screwing indentions.
> 
Hehe - maybe your editor is trying to be "smart"?:)
(git diff 66536883cee86001b237fd51a35bf76ac390db25 reveals that it
isn't;) )

> > - the global module "multicount" variable looks really fishy;)
> 
> I don't like it either. But I could not think of a better way. If you have an idea I would be thankful. 
> 
Ok. I'll try to have a look at it tomorrow.
Btw., thanks for the patch!

marcus-h commented:
On 2016-11-25 22:35:12 +0100, Marcus Hüwe wrote:
> (git diff 66536883cee86001b237fd51a35bf76ac390db25 reveals that it
> isn't;) )
> 
git show of course:)

adrianschroeter commented:
On Freitag, 25. November 2016, 08:21:53 CET wrote Marcus Hüwe:
> On 2016-11-25 07:43:27 -0800, Adrian Schröter wrote:
> > On Freitag, 25. November 2016, 07:26:18 CET wrote Marcus H�we:
...
> > But 
> > 
> >  osc r $YOUR_project kernel:kernel-source
> > 
> > would work (assuming kernel "kernel" as main container and "kernel-source.spec").
> > 
> > You can find an example setup here:
> > 
> >  https://build-test.opensuse.org/project/monitor/home:adrianSuSE
> > 
> Great, thanks!
> 
> Btw., https://build-test.opensuse.org states "Every openSUSE user can login",
> but it doesn't work for me:) (I'm getting a 401 using the same credentials
> as for api.o.o) (but the /public/ route is enough for this;) ).

right, I have improved the message and approved your account also for that instance.

Have a nice weekend!
adrian

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

marcus-h commented:
Does the attached patch work for you? (note: I renamed the "--mpackage"
option to "--subpackage", but I'm not really sure if this is a bit better
name...).
Btw., we should probably only allow the --mpackage/--subpackage
option, if CWD is package working copy or if osc was invoked via
"osc results <project> <package>".

A bit feedback about your original patch:

```
+def show_results_meta(apiurl, prj, package=None, lastbuild=None, repository=[], arch=[], oldstate=None, multibuild=None, locallink=None, mpackage=None):
```
mpackage is not used in this function. Just pop of the "mpackage" key,
if present, from the kwargs dict in get_results.

```
@@ -5456,11 +5461,11 @@ def show_prj_results_meta(apiurl, prj):
     f = http_GET(u)
     return f.readlines()

-
 def result_xml_to_dicts(xml):
```
No need to remove the line:) (actually, pep8 "dictates" to have 2
empty lines between two module level functions (unfortunately, we
really enforce it so far...))

```
-            elif res['code'] in ('scheduled', ) and res['details']:
-                # highlight scheduled jobs with possible dispatch problems
-                res['status'] += '*'
```
Hmm is there a reason why you dropped this?
[results_mb.txt](https://github.com/openSUSE/osc/files/615041/results_mb.txt)

lethliel commented:
> Does the attached patch work for you?

Yes. The attached patch works perfectly. 👍 

> Hmm is there a reason why you dropped this?

No. It was not by intention. Just a copy and paste error. 

Thank you again for your comments and your explanation. It helps me a lot. (also for future contributions)

lethliel commented:
@marcus-h I applied your patch and rebased. Can you please review again?

lethliel commented:
changed **-s** to **-S** to reuse it in other commands. -s is already used in remotebuildlog
changed **--multi-build** to **--all-subpackages**. That goes better with the --subpackage option.

adrianschroeter commented:
On Freitag, 2. Dezember 2016, 06:29:11 CET wrote Marco Strigl:
> changed **-s** to **-S** to reuse it in other commands. -s is already used in remotebuildlog
> changed **--multi-build** to **--all-subpackages**. That goes better with the --subpackage option. 

hm, I understand subpackages as rpm binary subpackages in first place.

are you feeling the same?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imend�rffer, Jane Smithard, Graham Norton, HRB 21284 (AG N�rnberg) 
 
Maxfeldstra�e 5                         
90409 N�rnberg 
Germany

marcus-h commented:
On 2016-12-02 02:22:43 -0800, Marco Strigl wrote:
> @marcus-h I applied your patch and rebased. Can you please review again?
> 
I just thought about it again... the "--subpackage" option is a
stupid/misleading name. I think we should go back to your "--mpackage"
or a "--multibuild-package" name. But this somewhat conflicts with the
"--multi-package" option...

I would propose the following:
"--multi-package" -> "--multibuild"
"--mpackage" -> "--multibuild-package"

What do you think?

lethliel commented:
> I would propose the following:
> "--multi-package" -> "--multibuild"
> "--mpackage" -> "--multibuild-package"

That sounds sane. 
But i would leave the short forms of '-m' and '-S'. 

I everyone agrees I will change it today in the evening.

marcus-h commented:
On 2016-12-02 06:48:30 -0800, Marco Strigl wrote:
> > I would propose the following:
> > "--multi-package" -> "--multibuild"
> > "--mpackage" -> "--multibuild-package"
> 
> That sounds sane. 
> But i would leave the short forms of '-m' and '-S'. 
> 
Hmm maybe use "-M" and "-p"? We more or less consistently
use "-m" for "--message". "-M" and "-p" are also already
used by some commands, but they don't have a "consistent"
meaning...

lethliel commented:
I changed the long forms to your suggestion. 
But i would like to keep the short forms. 

**-p** is already used for prefered packages or packages which could be confusing using the multibuild feature.

marcus-h commented:
On 2016-12-04 05:23:02 -0800, Marco Strigl wrote:
> I changed the long forms to your suggestion. 
> But i would like to keep the short forms. 
> 
> **-p** is already used for prefered packages or packages which could be confusing using the multibuild feature.
> 
Hmm "-S" is used as well and there's (IMHO) no connotation to
"--multibuild-package", but well... I can live with "-m" and "-S", too:)

@adrianschroeter: any opinion?

Apart from that, we should adjust the descriptions and local variable
names etc., too (the subpackage stuff).
Btw., in commit 8e54a1b31624f4c620767df6509073b18017813b the commandline
module is broken.

lethliel commented:
> Btw., in commit 8e54a1b the commandline
> module is broken.

ups. fixed that. 

> Apart from that, we should adjust the descriptions and local variable
> names etc., too (the subpackage stuff).

I will wait for the discussion with option (short and long) names coming to an end. IMHO it makes no sense to make a rebase after every comment ;-) Plus it adds the possibility of errors (as in 8e54a1b)

lethliel commented:
ok another proposal: 

-m (--multibuild) for all packages
-M <pack> (--multibuild-package <pack>) for a specific subpackage

@adrianschroeter @marcus-h Is this ok for you?

adrianschroeter commented:
Sounds fine for me

(I don't think that someone will try to send a message when just displaying results, so I don't see the conflict with --message as mentioned by Marcus)

marcus-h commented:
On 2016-12-13 00:29:33 -0800, Marco Strigl wrote:
> ok another proposal: 
> 
> -m (--multibuild) for all packages
> -M <pack> (--multibuild-package <pack>) for a specific subpackage
> 
> @adrianschroeter @marcus-h Is this ok for you? 
> 
Ok. Go for it:)

lethliel commented:
I changed the parameters and also modified the variable names to fit the parameters.

marcus-h commented:
On 2016-12-13 07:31:05 -0800, Marco Strigl wrote:
> I changed the parameters and also modified the variable names to fit the parameters. 
> 
Merged. Thanks a lot for the PR and your patience!

lethliel commented:
> Merged. Thanks a lot for the PR and your patience!

No problem. Thanks for the feedback. A good way to learn.
